### PR TITLE
add shape-type metadata to geo_shape's doc-value

### PR DESCRIPTION
**Pull request by talevy**
This commit serializes the ShapeType of the indexed
geometry. The ShapeType can be useful for other future
features. For one thing: #49887 depends on the ability
to determine what the highest dimensional shape is
for centroid calculations.

Since GeometryCollection is opaque, an additional 
piece of data about the highest-dimension contained
within it is serialized and resolved at the time of reading.
This overhead is only present when indexing GeometryCollection.

relates #49569, #37206.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/Extent.java [hunk 1]
```diff
@@ -79,8 +77,6 @@ public void reset(int top, int bottom, int negLeft, int negRight, int posLeft, i
      * @param topRightY   the top-right y-coordinate
      */
     public void addRectangle(int bottomLeftX, int bottomLeftY, int topRightX, int topRightY) {
-        assert bottomLeftX <= topRightX;
-        assert bottomLeftY <= topRightY;
         this.bottom = Math.min(this.bottom, bottomLeftY);
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/Extent.java]: I think this assertions are actually important, why did you remove them?

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java [hunk 2]
```diff
@@ -53,12 +54,35 @@ public void reset(BytesRef bytesRef) throws IOException {
...
+            case LINEARRING:
+            case MULTILINESTRING:
+                return 2;
+            case POLYGON:
+            case MULTIPOLYGON:
+            case ENVELOPE:
+            case CIRCLE:
+                return 3;
+            case GEOMETRYCOLLECTION:
+                return input.readVInt();
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: Maybe we do not need an extra byte here, maybe we can have three different Geometry collections depending on the highest dimension?

File: server/src/main/java/org/elasticsearch/common/geo/Extent.java (hunk 1 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/Extent.java]: the geocentroid aggregator tests were failing, maybe I was too quick to do this and it is a problem with the test.

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java (hunk 2 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: are you suggesting there be GeometryCollection, GeometryCollection3, Geometryacollection2, Geometryacollection1   in ShapeType? Or should there be a new enum specific to the tree's types

File: server/src/main/java/org/elasticsearch/common/geo/Extent.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/Extent.java]: I think the extent will only work if this assertions are true

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java (hunk 2 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: Something like that, I haven't made up my mind. I do not want to overcomplicate things either.

Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I guess the question is which worlds are more complicated:

- extend ShapeType to include specific GeometryCollection types with additional information about the highest dimension. If no dimension specified, then it is unknown.
  - **cons:** This would require custom mapping for the hypothetical future `geo_shape_type` aggregation, which would would only be slightly ugly, but contained. It would also convolute the ShapeType to not map 1-to-1 to proper shapes
- create a new Enum dedicated for type.
  - **cons:** need switch statements on reading and writing for conversion
- keep the original ShapeType, but write an extra bit for the dimension
  - **cons:** takes an extra bit and extra CPU to read the extra dimensional information

Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I think option 2 is the clearer one of all the options. Since this type is dedicated to centroid support, it will feel cleaner if the CentroidCalculator owns tracking this information since your latest change updated it to be a geometry-visitor, so it can get this information too

Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: In reality we only need the highest dimension of the shape, the type is an `extra` :)

I think is more accurate if we keep dimension 0 for points, 1 for lines and two for polygons.

Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: > I think is more accurate if we keep dimension 0 for points, 1 for lines and two for polygons.

agreed. after I started updating it today, I felt the same way 👍 

> In reality we only need the highest dimension of the shape, the type is an extra :)

I think if we're already going to keep this information, we might as well group the two 
so that we can enable the shape-type aggregations we spoke of.

**Comment by talevy** [comment]
@iverase I tried introducing a new `DimensionalShapeType` that represents both the shape and the dimension. I do think it is uglier, but it does save that extra byte. let me know what you think.

What is not included -> a conversion from DimensionalShapeType -> ShapeType. This would only be useful when implementing #49569

**Comment by talevy** [comment]
run elasticsearch-ci/bwc
run elasticsearch-ci/default-distro

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java [hunk 3]
```diff
@@ -58,8 +58,7 @@ public void reset(BytesRef bytesRef) throws IOException {
      */
     public Extent getExtent() {
         if (treeOffset == 0) {
-            // TODO: Compress serialization of extent
-            input.setPosition(extentOffset);
+            getDimensionalShapeType(); // skip DimensionalShapeType
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I will add the dimensional type as the first byte of the serialise tree and skip it together with the centroid.

**Review by iverase — CHANGES_REQUESTED** [review]
In general looks good. I just think the dimensional type should be part of the centroid and should be skipped when getting the extent.

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java (hunk 3 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I left this because of the variable encoding of the enum. in order to properly skip it, it must first be read and cached. is that preferred? or is changing the encoding to be a byte better? since it is known to be a small amount of constants, it could be serialized to a byte with tests making sure any future changes do not break this

Comment (iverase) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I would expect the enum to be one byte, ordinal from 1 to 7 must fit in a byte

Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I agree, but I'd be happier to be explicit about this, instead of just trusting that vint will return a byte. I'll update!

**Comment by talevy** [comment]
run elasticsearch-ci/1

**Comment by talevy** [comment]
run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java [hunk 4]
```diff
@@ -33,8 +33,8 @@
  * relations against the serialized triangle tree.
  */
 public class TriangleTreeReader {
+    private static final int CENTROID_HEADER_SIZE_IN_BYTES = 9;
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: Maybe add a comment where this 9 bytes are coming from.

**Review by iverase — APPROVED** [review]
Just left a small comment. LGTM

File: server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java (hunk 4 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java]: I will update this in the follow-up PR when this is modified to include the centroid weight!

**Comment by talevy** [comment]
thanks Ignacio!
